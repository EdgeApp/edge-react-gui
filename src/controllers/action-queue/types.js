// @flow
/* eslint-disable no-use-before-define */

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'exchange-buy'
  | 'exchange-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type ActionOpExecStatus = 'pending' | 'active' | 'done' | Error

export type ActionOp =
  | {
      type: 'seq',
      actions: ActionOp[]
    }
  | {
      type: 'par',
      actions: ActionOp[]
    }
  | {
      type: 'broadcast-tx',
      pluginId: string,
      rawTx: Uint8Array
    }
  | {
      type: 'exchange-buy',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'exchange-sell',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'loan-borrow',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-deposit',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-repay',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-withdraw',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'swap',
      fromWalletId: string,
      toWalletId: string,
      fromTokenId?: string,
      toTokenId?: string,
      nativeAmount: string,
      amountFor: 'from' | 'to'
    }
  // Useful for development/testing
  | {
      type: 'toast',
      message: string
    }
  | {
      type: 'delay',
      ms: number
    }

//
// Action (After) Effects
//

export type ActionEffect =
  | {
      type: 'seq',
      opIndex: number,
      childEffect: ActionEffect
    }
  | {
      type: 'par',
      childEffects: ActionEffect[]
    }
  | {
      type: 'address-balance',
      address: string,
      aboveAmount?: string,
      belowAmount?: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'tx-confs',
      txId: string,
      walletId: string,
      confirmations: number
    }
  | {
      type: 'price-level',
      currencyPair: string,
      aboveRate?: number,
      belowRate?: number
    }
  | {
      type: 'done',
      error?: Error
    }
  // Useful for development/testing
  | {
      type: 'unixtime',
      timestamp: number
    }
  | {
      type: 'init'
    }

//
// Action Program
//

// On Disk:
export type ActionProgram = {
  programId: string,
  actionOp: ActionOp
}

export type ActionProgramState = {
  programId: string,
  effect?: ActionEffect
}

//
// Internal Types
//

export type ActionQueueItem = {
  program: ActionProgram,
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

export type ExecutionResult = {
  effect: ActionEffect,
  action?: ActionOp // For dryrun
}
export type ExecutionResults = {
  nextState: ActionProgramState
}

//
// Aciton Display API
//

export type ActionDisplayInfo = {
  title: string,
  message: string,
  status: ActionOpExecStatus,
  steps: ActionDisplayInfo[]
}
