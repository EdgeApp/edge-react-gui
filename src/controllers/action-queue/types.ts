/* eslint-disable no-use-before-define */

import { EdgeAccount, EdgeNetworkFee, EdgeTransaction } from 'edge-core-js'

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'wyre-buy'
  | 'wyre-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type SeqActionOp = {
  type: 'seq'
  actions: ActionOp[]
}
export type ParActionOp = {
  type: 'par'
  actions: ActionOp[]
}
export type BroadcastTxActionOp = {
  type: 'broadcast-tx'
  pluginId: string
  rawTx: Uint8Array
}
export type WyreBuyActionOp = {
  type: 'wyre-buy'
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type WyreSellActionOp = {
  type: 'wyre-sell'
  wyreAccountId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanBorrowActionOp = {
  type: 'loan-borrow'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanDepositActionOp = {
  type: 'loan-deposit'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanRepayActionOp = {
  type: 'loan-repay'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanWithdrawActionOp = {
  type: 'loan-withdraw'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type SwapActionOp = {
  type: 'swap'
  amountFor: 'from' | 'to'
  fromTokenId?: string
  fromWalletId: string
  nativeAmount: string
  toTokenId?: string
  toWalletId: string
}
export type ActionOp =
  | SeqActionOp
  | ParActionOp
  | BroadcastTxActionOp
  | WyreBuyActionOp
  | WyreSellActionOp
  | LoanBorrowActionOp
  | LoanDepositActionOp
  | LoanRepayActionOp
  | LoanWithdrawActionOp
  | SwapActionOp

//
// Action (After) Effects
//

export type SeqEffect = {
  type: 'seq'
  opIndex: number
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type ParEffect = {
  type: 'par'
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type AddressBalanceEffect = {
  type: 'address-balance'
  address: string
  aboveAmount?: string
  belowAmount?: string
  walletId: string
  tokenId?: string
}
export type PushEventEffect = {
  type: 'push-event'
  eventId: string
}
export type PriceLevelEffect = {
  type: 'price-level'
  currencyPair: string
  aboveRate?: number
  belowRate?: number
}
export type TxConfsEffect = {
  type: 'tx-confs'
  txId: string
  walletId: string
  confirmations: number
}
export type DoneEffect = {
  type: 'done'
  error?: Error
  cancelled?: boolean
}

export type ActionEffect = SeqEffect | ParEffect | AddressBalanceEffect | PushEventEffect | PriceLevelEffect | TxConfsEffect | DoneEffect

//
// Action Program
//

// Storage:
export type ActionProgram = {
  programId: string
  actionOp: ActionOp
  // Development mode flag
  mockMode?: boolean
}
export type ActionProgramState = {
  clientId: string
  programId: string
  effect?: ActionEffect

  // Flags:
  effective: boolean // Whether the effect is observed
  executing: boolean // Whether the program is executing

  lastExecutionTime: number // The time when the effect was checked
  nextExecutionTime: number // The next time when the effect should be checked again
}

export type ActionQueueItem = {
  program: ActionProgram
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

// Runtime:
export type BroadcastTx = {
  walletId: string
  networkFee: EdgeNetworkFee
  tx: EdgeTransaction
}
export type ExecutableAction = {
  dryrunOutput: ExecutionOutput | null
  execute: () => Promise<ExecutionOutput>
}
export type ExecutionContext = {
  account: EdgeAccount
  clientId: string
}
export type ExecutionOutput = {
  effect: ActionEffect
  broadcastTxs: BroadcastTx[]
}
export type ExecutionResults = {
  nextState: ActionProgramState
}
export type PendingTxMap = {
  [walletId: string]: EdgeTransaction[]
}

//
// Aciton Display API
//

export type ActionDisplayStatus = 'pending' | 'active' | 'done' | Error

export type ActionDisplayInfo = {
  title: string
  message: string
  status: ActionDisplayStatus
  steps: ActionDisplayInfo[]
}
