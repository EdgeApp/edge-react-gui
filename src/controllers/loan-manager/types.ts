import { BorrowEngine, BorrowPlugin } from '../../plugins/borrow-plugins/types'
import { LoanProgramEdge } from './store'

export type LoanAccount = {
  id: string
  borrowPlugin: BorrowPlugin
  borrowEngine: BorrowEngine

  // Action Program Relationships
  programEdges: LoanProgramEdge[]
}

export type LoanAccountMap = {
  [walletId: string]: LoanAccount
}

export type LoanManager = {
  loanAccounts: LoanAccount[]
}
