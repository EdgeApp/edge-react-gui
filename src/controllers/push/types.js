// @flow

// -----------------------------------------------------------------------------
// Push Server Types
// -----------------------------------------------------------------------------

export type Device = {
  appId: string,
  deviceId: string,
  deviceToken: string,
  rootLoginIds: Uint8Array[],
  created: Date,
  visited: Date
}

//
// Push Triggers
//

export type PriceChangeTrigger = {
  type: 'price-change',
  pluginId: string,
  tokenId: string,
  aboveRatio?: number,
  belowRatio?: number,
  range: 'hourly' | 'daily'
}

export type PriceLevelTrigger = {
  type: 'price-level',
  currencyPair: string,
  aboveRate?: number,
  belowRate?: number
}

export type AddressBalanceTrigger = {
  type: 'address-balance',
  pluginId: string,
  tokenId?: string,
  address: string,
  aboveAmount?: string,
  belowAmount?: string
}

export type TxConfirmTrigger = {
  type: 'tx-confirm',
  pluginId: string,
  confirmations: number,
  txid: string
}

export type PushTrigger = AddressBalanceTrigger | PriceChangeTrigger | PriceLevelTrigger | TxConfirmTrigger

//
// Push Event
//

export type PushEvent = {
  trigger: PushTrigger,
  triggered: boolean,

  pushMessage?: string,
  pushPayload?: mixed,
  broadcastTxs?: Array<{
    pluginId: string,
    rawTx: Uint8Array
  }>
}

//
// HTTP
//

export type PushRequestBody = {
  // The request payload:
  data: mixed,

  // Who is making the request:
  apiKey: string,
  appId: string,
  deviceId: string,

  // For logins:
  rootLoginId?: Uint8Array
  // rootSecretHash?: Uint8Array
}

/**
 * Refreshes the `visited` date on a device and its effects or logins.
 * Devices expire after 3 months.
 * POST /v2/device/checkin
 */

/**
 * Registers / updates a device.
 * POST /v2/device/update
 */
export interface UpdateDevicePayload {
  rootLoginIds: Uint8Array[]; // asArray(asBase64)
  events: PushEvent[];
  deviceToken: string;
}

/**
 * Gets a device from the database.
 * POST /v2/device
 */
export type DevicePayload = {
  deviceId: string,
  deviceToken: string,
  events: PushEvent[],
  rootLoginIds: Uint8Array[], // asArray(asBase64)
  created: Date,
  visited: Date
}

/**
 * Registers / updates a login.
 * POST /v2/login/update
 */
export type UpdateLoginPayload = {
  events: PushEvent[]

  // removeEvents?: string[]
  // replaceEvents?: PushEvent[]
  // newEvents?: PushEvent[]
}

/**
 * Reads a login from the database.
 * POST /v2/login
 */
export type LoginPayload = {
  deviceIds: string,
  events: PushEvent[],
  created: Date,
  visited: Date
}
