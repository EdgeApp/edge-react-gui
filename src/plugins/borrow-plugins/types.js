// @flow

import { type EdgeCurrencyWallet, type EdgeNetworkFee, type EdgeTransaction } from 'edge-core-js'

// -----------------------------------------------------------------------------
// Request Method Types
// -----------------------------------------------------------------------------

// Borrow more:
export type BorrowRequest = {
  tokenId?: string,
  nativeAmount: string,

  // Optional source for the funds which will borrow on behalf of the borrow engine's currencyWallet
  fromWallet?: EdgeCurrencyWallet
}

// Make payment:
export type RepayRequest = {
  tokenId?: string,
  nativeAmount: string,

  // Optional source for the funds which will repay on behalf of the borrow engine's currencyWallet
  fromWallet?: EdgeCurrencyWallet
}

// Deposit collateral:
export type DepositRequest = {
  tokenId?: string,
  nativeAmount: string,

  // Optional source for the funds which will deposit on behalf of the borrow engine's currencyWallet
  fromWallet?: EdgeCurrencyWallet
}

// Withdraw collateral:
export type WithdrawRequest = {
  tokenId?: string,
  nativeAmount: string,

  // Optional destination for the funds
  toWallet?: EdgeCurrencyWallet
}

// General purpose approvable action
export type ApprovableAction = {
  +networkFee: EdgeNetworkFee,
  +unsignedTxs: EdgeTransaction[],
  +approve: () => Promise<EdgeTransaction[]>
}

// HACK: Used to identify running ActionQueue programs by borrow action type.
export type BorrowActionId = 'loan-create' | 'loan-deposit' | 'loan-borrow' | 'loan-repay' | 'loan-withdraw'

// -----------------------------------------------------------------------------
// Engine
// -----------------------------------------------------------------------------

export type BorrowCollateral = {
  tokenId?: string,
  nativeAmount: string
}

export type BorrowDebt = {
  tokenId?: string,
  nativeAmount: string,
  apr: number
}

export type BorrowEngine = {
  // Currency wallet for the loan account (e.g. Ethereum wallet)
  currencyWallet: EdgeCurrencyWallet,

  collaterals: BorrowCollateral[],
  debts: BorrowDebt[],

  loanToValue: number,

  // Returns the APR for borrow a particular token
  getAprQuote: (tokenId?: string) => Promise<number>,

  // Collateral modification
  deposit: (request: DepositRequest) => Promise<ApprovableAction>,
  withdraw: (request: WithdrawRequest) => Promise<ApprovableAction>,

  // Debt modification
  borrow: (request: BorrowRequest) => Promise<ApprovableAction>,
  repay: (request: RepayRequest) => Promise<ApprovableAction>,

  // Close loan account
  close: () => Promise<ApprovableAction>
}

// -----------------------------------------------------------------------------
// Plugin Info
// -----------------------------------------------------------------------------

export type BorrowPluginInfo = {
  pluginId: string,
  displayName: string,

  // Defines the relationship to the type of currency plugins which the plugin supports
  currencyPluginId: string,

  // Maximum loan-to-value ratio
  maxLtvRatio: number
}

// -----------------------------------------------------------------------------
// Plugin
// -----------------------------------------------------------------------------

export type BorrowPlugin = {
  borrowInfo: BorrowPluginInfo,
  makeBorrowEngine: (wallet: EdgeCurrencyWallet) => Promise<BorrowEngine>
}
