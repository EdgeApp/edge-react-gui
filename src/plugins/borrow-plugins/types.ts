import { EdgeCurrencyWallet, EdgeNetworkFee, EdgeTransaction } from 'edge-core-js'
import { Subscriber } from 'yaob'

import { PendingTxMap } from '../../controllers/action-queue/types'

// -----------------------------------------------------------------------------
// Request Method Types
// -----------------------------------------------------------------------------

// Borrow more:
export type BorrowRequest = {
  tokenId?: string
  nativeAmount: string

  // Optional source for the funds which will borrow on behalf of the borrow engine's currencyWallet
  fromWallet?: EdgeCurrencyWallet
}

// Make payment:
export type RepayRequest = {
  tokenId?: string
  nativeAmount: string

  // Optional source for the funds which will repay on behalf of the borrow engine's currencyWallet
  fromWallet?: EdgeCurrencyWallet
  fromTokenId?: string
}

// Deposit collateral:
export type DepositRequest = {
  tokenId?: string
  nativeAmount: string

  // Optional source for the funds which will deposit on behalf of the borrow engine's currencyWallet
  fromWallet?: EdgeCurrencyWallet
}

// Withdraw collateral:
export type WithdrawRequest = {
  tokenId?: string
  nativeAmount?: string

  // Optional destination for the funds
  toWallet?: EdgeCurrencyWallet
}

export type BroadcastTx = {
  walletId: string
  networkFee: EdgeNetworkFee
  tx: EdgeTransaction
}

// General purpose approvable action
export type ApprovableAction = {
  readonly networkFee: EdgeNetworkFee
  readonly unsignedTxs: EdgeTransaction[]
  // Optional pending txs to pass along to the wallet when making transactions
  readonly dryrun: (pendingTxMap: PendingTxMap) => Promise<BroadcastTx[]>
  readonly approve: () => Promise<BroadcastTx[]>
}

// HACK: Used to identify running ActionQueue programs by borrow action type.
export type BorrowActionId = 'loan-create' | 'loan-deposit' | 'loan-borrow' | 'loan-repay' | 'loan-withdraw'

// -----------------------------------------------------------------------------
// Engine
// -----------------------------------------------------------------------------

export type BorrowCollateral = {
  tokenId?: string
  nativeAmount: string
}

export type BorrowDebt = {
  tokenId?: string
  nativeAmount: string
  apr: number
}

export type BorrowEngine = {
  // Currency wallet for the loan account (e.g. Ethereum wallet)
  currencyWallet: EdgeCurrencyWallet

  collaterals: BorrowCollateral[]
  debts: BorrowDebt[]

  loanToValue: number

  readonly watch: Subscriber<BorrowEngine>
  syncRatio: number

  // Returns the APR for borrow a particular token
  getAprQuote: (tokenId?: string) => Promise<number>

  // Collateral modification
  deposit: (request: DepositRequest) => Promise<ApprovableAction>
  withdraw: (request: WithdrawRequest) => Promise<ApprovableAction>

  // Debt modification
  borrow: (request: BorrowRequest) => Promise<ApprovableAction>
  repay: (request: RepayRequest) => Promise<ApprovableAction>

  // Close loan account
  close: () => Promise<ApprovableAction>
}

// -----------------------------------------------------------------------------
// Plugin Info
// -----------------------------------------------------------------------------

export type BorrowPluginInfo = {
  borrowPluginId: string

  // Display information
  displayName: string
  // A token from the currencyPluginId, for displaying as the Loan provider icon
  displayTokenId: string

  // Defines the relationship to the of currency plugins which the plugin supports
  currencyPluginId: string

  // Maximum loan-to-value ratio
  maxLtvRatio: number
}

// -----------------------------------------------------------------------------
// Plugin
// -----------------------------------------------------------------------------

export type BorrowPlugin = {
  borrowInfo: BorrowPluginInfo
  makeBorrowEngine: (wallet: EdgeCurrencyWallet) => Promise<BorrowEngine>
}
