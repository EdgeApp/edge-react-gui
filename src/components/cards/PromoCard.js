// @flow

import * as React from 'react'
import { TouchableOpacity } from 'react-native'
import FastImage from 'react-native-fast-image'
import { TouchableWithoutFeedback } from 'react-native-gesture-handler'
import Animated, { interpolate, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'
import AntDesignIcon from 'react-native-vector-icons/AntDesign'

import { useHandler } from '../../hooks/useHandler.js'
import { type Theme, cacheStyles, getTheme } from '../services/ThemeContext.js'
import { EdgeText } from '../themed/EdgeText.js'

type Props = {
  message: string,
  iconUri?: string,
  animationValue: Animated.SharedValue<number>,
  onPress: () => void,
  onClose: () => void
}

export const PromoCard = (props: Props) => {
  const { message, iconUri, onPress, onClose, animationValue } = props
  const theme = getTheme()
  const styles = getStyles(theme)

  const pressed = useSharedValue(0)

  const handlePressIn = useHandler(() => (pressed.value = 1))
  const handlePressOut = useHandler(() => (pressed.value = 0))

  const borderAnimatedStyle = useAnimatedStyle(() => {
    return {
      borderColor: withTiming(pressed.value ? theme.promoCardBorderColorPressIn : theme.promoCardBorderColorPressOut, { duration: 100 }),
      opacity: interpolate(animationValue.value, [-1, 0, 1], [0.5, 1, 0.5])
    }
  })

  const contentAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: withTiming(interpolate(animationValue.value, [-1, 0, 1], [0, 1, 0]), { duration: 10 })
    }
  })

  return (
    <TouchableWithoutFeedback onPressIn={handlePressIn} onPressOut={handlePressOut} activeOpacity={1} onPress={onPress}>
      <Animated.View style={[styles.container, borderAnimatedStyle]}>
        <Animated.View style={[styles.content, contentAnimatedStyle]}>
          {iconUri != null ? <FastImage resizeMode="contain" source={{ uri: iconUri }} style={styles.icon} /> : null}

          <EdgeText numberOfLines={0} style={styles.text}>
            {message}
          </EdgeText>

          <TouchableOpacity onPress={onClose}>
            <AntDesignIcon name="close" color={theme.iconTappable} size={theme.rem(2.5)} style={styles.close} />
          </TouchableOpacity>
        </Animated.View>
      </Animated.View>
    </TouchableWithoutFeedback>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    backgroundColor: theme.promoCardBackground,
    borderWidth: theme.promoCardBorderWidth,
    borderRadius: theme.rem(0.75),
    padding: theme.rem(1),
    height: '100%'
  },
  content: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center'
  },
  icon: {
    width: theme.rem(4),
    height: theme.rem(4)
  },
  text: {
    flex: 1,
    marginLeft: theme.rem(1),
    fontSize: theme.rem(3.5)
  },
  close: {}
}))
