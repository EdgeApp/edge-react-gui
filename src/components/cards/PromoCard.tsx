import * as React from 'react'
import { TouchableOpacity, View } from 'react-native'
import FastImage from 'react-native-fast-image'
import AntDesignIcon from 'react-native-vector-icons/AntDesign'

import { hideMessageTweak } from '../../actions/AccountReferralActions'
import { linkReferralWithCurrencies } from '../../actions/WalletListActions'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { bestOfMessages } from '../../util/ReferralHelpers'
import { showError } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { ButtonBox } from '../themed/ThemedButtons'

interface Props {
  navigation: NavigationBase
}

export function PromoCard(props: Props) {
  const { navigation } = props
  const theme = useTheme()
  const styles = getStyles(theme)
  const dispatch = useDispatch()

  const accountMessages = useSelector(state => state.account.referralCache.accountMessages)
  const accountReferral = useSelector(state => state.account.accountReferral)
  const messageSummary = bestOfMessages(accountMessages, accountReferral)

  const handlePress = useHandler(() => {
    const uri = messageSummary?.message.uri
    if (uri != null) dispatch(linkReferralWithCurrencies(navigation, uri)).catch(err => showError(err))
  })

  const handleClose = useHandler(() => {
    if (messageSummary != null) {
      dispatch(hideMessageTweak(messageSummary.messageId, messageSummary.messageSource)).catch(err => showError(err))
    }
  })

  if (messageSummary == null) return null
  const { message } = messageSummary
  return (
    <ButtonBox marginRem={1} onPress={handlePress}>
      <View style={styles.container}>
        {message.iconUri != null ? <FastImage resizeMode="contain" source={{ uri: message.iconUri }} style={styles.icon} /> : null}
        <EdgeText testID="promoCard" numberOfLines={0} style={styles.text}>
          {message.message}
        </EdgeText>
        <TouchableOpacity accessible={false} onPress={handleClose}>
          <AntDesignIcon
            testID="closePromo"
            name="close"
            color={theme.iconTappable}
            size={theme.rem(1)}
            style={styles.close}
            accessibilityHint={lstrings.close_hint}
          />
        </TouchableOpacity>
      </View>
    </ButtonBox>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: theme.tileBackground,
    padding: theme.rem(0.5)
  },
  icon: {
    width: theme.rem(2),
    height: theme.rem(2),
    margin: theme.rem(0.5)
  },
  text: {
    flex: 1,
    margin: theme.rem(0.5)
  },
  close: {
    padding: theme.rem(0.5)
  }
}))
