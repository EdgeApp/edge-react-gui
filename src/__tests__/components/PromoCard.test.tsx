import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import renderer from 'react-test-renderer'

import { PromoCard } from '../../components/cards/PromoCard'
import { MessageTweak } from '../../types/TweakTypes'
import { fakeNavigation } from '../../util/fake/fakeNavigation'
import { FakeProviders, FakeState } from '../../util/fake/FakeProviders'

describe('PromoCard', () => {
  const fakeMessage: MessageTweak = {
    message: 'Show me',
    iconUri: 'https://edge.app/favicon.ico',
    durationDays: 1
  }

  const mockState: FakeState = {
    account: {
      accountReferral: {
        installerId: 'string',
        currencyCodes: ['BTC'],
        promotions: [],
        ignoreAccountSwap: true,
        hiddenAccountMessages: { messageId: true }
      },
      referralCache: {
        accountMessages: [fakeMessage]
      }
    }
  }

  it('should render', () => {
    const actual = renderer.create(
      <FakeProviders initialState={mockState}>
        <PromoCard navigation={fakeNavigation} />
      </FakeProviders>
    )

    expect(actual).toMatchSnapshot()
  })
})
