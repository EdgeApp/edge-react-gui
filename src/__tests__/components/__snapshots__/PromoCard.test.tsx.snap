// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromoCard should render 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={false}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "marginBottom": 22,
      "marginLeft": 22,
      "marginRight": 22,
      "marginTop": 22,
      "opacity": 1,
      "paddingBottom": 11,
      "paddingLeft": 11,
      "paddingRight": 11,
      "paddingTop": 11,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "flexDirection": "row",
        "padding": 11,
      }
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "height": 45,
            "margin": 11,
            "width": 45,
          },
        ]
      }
    >
      <FastImageView
        resizeMode="contain"
        source={
          {
            "uri": "https://edge.app/favicon.ico",
          }
        }
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
    <Text
      adjustsFontSizeToFit={true}
      minimumFontScale={0.65}
      numberOfLines={0}
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Quicksand-Regular",
            "fontSize": 22,
            "includeFontPadding": false,
          },
          {
            "flex": 1,
            "margin": 11,
          },
          null,
        ]
      }
      testID="promoCard"
    >
      Show me
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={false}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        accessibilityHint="Close"
        allowFontScaling={false}
        style={
          [
            {
              "color": "#00f1a2",
              "fontSize": 22,
            },
            {
              "padding": 11,
            },
            {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
        testID="closePromo"
      >
        î˜³
      </Text>
    </View>
  </View>
</View>
`;
